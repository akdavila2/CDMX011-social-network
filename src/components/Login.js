/* eslint-disable eol-last */
/* eslint-disable semi */
/* eslint-disable space-before-function-paren */
/* eslint-disable indent */
/* eslint-disable import/no-cycle */
import { login, loginGoogle } from '../lib/firebase.js';
import { onNavigate } from '../router/router.js';

export const Login = () => {
    const view = `
<div class="acount-header">
    <img src="../img/logoFormLoveBook.png" class="acount-logo" alt="LoveBook logo">
</div>
<div class="log-content">
    <form action="" id="login-form" class="form">
        <label for="login-email">Email</label>
        <input type="email" id="loginEmail" required>
        <label for="login-password">Password</label>
        <input type="password" id="loginPassword" required>
        <div class="error-message" id="loginMessages"></div>
        <button type="submit" id="btnLogin">Login</button>
        <div class="division">
            <hr class="division-1">
            <p class="division-2">O</p>
            <hr class="division-3">
        </div>
        <div class="login-google">
            <img src="../img/icongoogle.png" alt="logo google" class="logo-google">
            <a id="withGoogle">Login with google</a>
        </div>
        <div class="pieform">
            <p style="font-size: 12px;">You do not have an account</p>   
        </div>
        <button type="submit" id="signup">Sign Up</button>
    </form>
</div>`;
    const loginContainer = document.createElement('div');
    loginContainer.classList.add('acount');
    loginContainer.innerHTML = view;
    const btnLogin = loginContainer.querySelector('#btnLogin');
    const btnSignUp = loginContainer.querySelector('#signup');
    const withGoogle = loginContainer.querySelector('#withGoogle');
    btnLogin.addEventListener('click', async(event) => {
        event.preventDefault();
        const email = loginContainer.querySelector('#loginEmail').value;
        const password = loginContainer.querySelector('#loginPassword').value;

        try {
            await login(email, password);
            // onNavigate('/home');
        } catch (error) {
            loginContainer.querySelector('#loginMessages').innerHTML = '&#x02716'.concat(' ', error.message);
        }
    });
    btnSignUp.addEventListener('click', (event) => {
        event.preventDefault();
        onNavigate('/signUp');
    });
    // Login con google
    withGoogle.addEventListener('click', async(event) => {
        event.preventDefault();
        try {
            await loginGoogle();
            onNavigate('/home');
        } catch (error) {
            loginContainer.querySelector('#loginMessages').innerHTML = '&#x02716'.concat(' ', error.message);
        }
    });
    return loginContainer;
};