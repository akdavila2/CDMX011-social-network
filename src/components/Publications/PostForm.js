import { onNavigate } from "../../router/router.js"
import { Navbar } from "../Navbar.js"
import { savePost } from "../../lib/firebase.js"

export const PostForm = () => {
    const viewInfo = `
<div id="addPublication">
    <div class= "form-Background">
    <form action="" id="publicationsForm" class="form">
        <label for="title">Title of the book</label>
        <input class = "inputpost" type="text" id="title">
        <label for="rating">Rating</label>
        <input class = "inputpost" type="number" id="rating" min= 1 max=10>
        <label for="review">Review</label>
        <textarea class = "inputpost" type="text" id="review"></textarea>
        <div class="error-message" id="postMesseges"></div>
        <div class= "text-publication">
             <a class="textaddpost" href="/home" id="btnAddPoster"><img class="icon-post" src="../img/plusazul.png"> Add Post</a>
        </div>
      
    </form>
    </div>
</div>
`
    const post = document.createElement('div');
    post.appendChild(Navbar());
    const contentPost = document.createElement('div');
    contentPost.classList.add('containerpost');
    contentPost.innerHTML = viewInfo;
    post.appendChild(contentPost);
    const btnAddPost = contentPost.querySelector('#btnAddPoster');
    const db = firebase.firestore();
    btnAddPost.addEventListener('submit', async(event) => {
        event.preventDefault();

        const title = contentPost.querySelector('#title').value;
        const rating = contentPost.querySelector('#rating').value;
        const review = contentPost.querySelector('#review').value;
        try {
            await savePost(title, rating, review)
            onNavigate('/home');
        } catch (error) {
            signupContainer.querySelector('#postMesseges').innerHTML = '&#x02716'.concat(' ', error.message)
        }


    });
    return post;
}