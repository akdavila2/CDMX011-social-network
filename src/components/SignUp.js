/* eslint-disable eol-last */
/* eslint-disable space-before-function-paren */
/* eslint-disable indent */
/* eslint-disable import/no-cycle */
import { onNavigate } from '../router/router.js';
import { register } from '../lib/firebase.js';

export const SignUp = () => {
    const view = `
        <div class="acount-header">
            <img src="../img/logoFormLoveBook.png" class="acount-logo" alt="LoveBook logo">
        </div>
        <div class="log-content">
            <form action="" id="signupForm" class="form">
                <label for="signupUserName">Username</label>
                <input type="text" id="signupUserName">
                <label for="signupEmail">Email</label>
                <input type="email" id="signupEmail" required>
                <label for="signupPassword">Password</label>
                <input type="password" id="signupPassword" required>
                <label for="signupPassword2">Confirm your password</label>
                <input type="password" id="signupPassword2" required>
                <div class="error-message" id="signupMesseges"></div>
                <button type="submit" id="btnSendSignUp">Sign Up</button>
            </form>
        </div>
        `;

    const signupContainer = document.createElement('div');
    signupContainer.classList.add('acount');

    signupContainer.innerHTML = view;

    const btnSendSignUp = signupContainer.querySelector('#btnSendSignUp');

    btnSendSignUp.addEventListener('click', async(event) => {
        event.preventDefault();

        const email = signupContainer.querySelector('#signupEmail').value;
        const password = signupContainer.querySelector('#signupPassword').value;
        const signUpPassword2 = signupContainer.querySelector('#signupPassword2').value;

        if (password === signUpPassword2) {
            try {
                await register(email, password);
                onNavigate('/home');
            } catch (error) {
                signupContainer.querySelector('#signupMesseges').innerHTML = '&#x02716'.concat(' ', error.message);
            }
        } else {
            signupContainer.querySelector('#signupMesseges').innerHTML = '&#x02716 Passwords do not match';
        }
    });
    return signupContainer;
};