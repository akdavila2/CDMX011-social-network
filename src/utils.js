/* eslint-disable semi */
/* eslint-disable eol-last */
/* eslint-disable indent */
export const render = (target, element) => {
    while (target.firstChild) {
        target.removeChild(target.lastChild);
    }
    target.append(element);
}