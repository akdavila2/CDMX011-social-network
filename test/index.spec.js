/**
 * @jest-environment jsdom
 */
/* eslint-disable space-before-function-paren */
/* eslint-disable eol-last */
/* eslint-disable indent */
import './globals/firebaseTest.js';
import { render } from '../src/utils.js';
import { Login } from '../src/components/Login.js';
import { SignUp } from '../src/components/SignUp.js';
import { PostForm } from '../src/components/Publications/PostForm.js';
// Login component tests
const delay = (timeInMs = 100) => new Promise((resolve) => setTimeout(resolve, timeInMs));
describe('Login', () => {
    document.body.innerHTML = '<div id="root"></div>';
    const rootDiv = document.getElementById('root');
    const loginForm = (email, password) => {
        rootDiv.querySelector('#loginEmail').value = email;
        rootDiv.querySelector('#loginPassword').value = password;
        rootDiv.querySelector('#btnLogin').click();
    };
    it('should be a function', () => {
        expect(typeof Login).toBe('function');
    });
    it('Should render the login screen', () => {
        render(rootDiv, Login());
        expect(rootDiv.innerHTML).toMatchSnapshot();
    });
    it('Should login user successfully', async() => {
        const email = 'test@labo.com';
        const password = '123456';
        const user = { email, uid: '42xtUMfM3cSZ61Arn4aH' }; // un objeto dummy que representa el user
        const mockSignInWithEmailAndPassword = jest.fn();
        mockSignInWithEmailAndPassword.mockResolvedValue(user);
        const mockFirebaseAuth = {
            signInWithEmailAndPassword: mockSignInWithEmailAndPassword,
            currentUser: user,
        };
        firebase.auth = () => mockFirebaseAuth;
        render(rootDiv, Login());
        loginForm(email, password);
        await delay(100);
        expect(mockSignInWithEmailAndPassword).toHaveBeenCalledWith(email, password);
    });
    it('shows an error message when credentials are invalid', async() => {
        const email = 'test@labo.com';
        const password = 'invalidpassword';
        const errorMessage = 'Invalid credentials!';
        const mockSignInWithEmailAndPassword = jest.fn();
        mockSignInWithEmailAndPassword.mockRejectedValue(new Error(errorMessage));

        const mockFirebaseAuth = {
            signInWithEmailAndPassword: mockSignInWithEmailAndPassword,
        };
        firebase.auth = () => mockFirebaseAuth;

        render(rootDiv, Login());
        loginForm(email, password);
        await delay(100);
        expect(mockSignInWithEmailAndPassword).toHaveBeenCalledWith(email, password);
        expect(rootDiv.innerHTML).toContain(errorMessage);
    });
    it('Expected event when clicking the signup button', () => {
        const btnSignUp = rootDiv.querySelector('#signup');
        btnSignUp.click();
        expect(btnSignUp.outerHTML).toBe('<button type="submit" id="signup">Sign Up</button>');
    });
});

// SignUp component tests
describe('SignUp', () => {
    document.body.innerHTML = '<div id="root"></div>';
    it('Should render SignUp', () => {
        const rootDiv = document.getElementById('root');
        const component = SignUp();
        render(rootDiv, component);
        expect(rootDiv.innerHTML).toMatchSnapshot();
    });
    it('Should Login the user when the SignUp button is clicked', () => {
        const mockSignUp = jest.fn();
        mockSignUp.mockImplementation(() => Promise.resolve());

        firebase.auth = jest.fn().mockImplementation(() => ({
            createUserWithEmailAndPassword: mockSignUp,
        }));

        const rootDiv = document.getElementById('root');
        const component = SignUp();
        render(rootDiv, component);

        const email = 'karina@labo.com';
        const password = '123456';
        const signUpPassword2 = '123456';

        document.querySelector('#signupEmail').value = email;
        document.querySelector('#signupPassword').value = password;
        document.querySelector('#signupPassword2').value = signUpPassword2;
        const btnSignUp = document.getElementById('btnSendSignUp');
        btnSignUp.click();
        expect(mockSignUp).toHaveBeenCalledWith(email, password);
    });
    it('Expected event when clicking the signup button', () => {
        const btnSignUp = document.getElementById('btnSendSignUp');
        btnSignUp.click();
        expect(btnSignUp.outerHTML).toBe('<button type="submit" id="btnSendSignUp">Sign Up</button>');
    });
});

// RendPosts component test
describe('PostForm', () => {
    document.body.innerHTML = '<div id="root"></div>';
    it('Should render', () => {
        const rootDiv = document.getElementById('root');
        const component = PostForm();
        render(rootDiv, component);
        expect(rootDiv.innerHTML).toMatchSnapshot();
    });
    it('Expected event when clicking the AddPosts button', () => {
        const btnAddPost = document.querySelector('#btnAddPoster');
        btnAddPost.click();
        expect(btnAddPost.outerHTML).toBe('<button class="btnPost" id="btnAddPoster" type="submit"><img class="icon-plus" src="../img/plusazul.png">Add Post</button>');
    });
});