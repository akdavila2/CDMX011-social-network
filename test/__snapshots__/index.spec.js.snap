// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Should render the login screen 1`] = `
"<div class=\\"acount\\">
<div class=\\"acount-header\\">
    <img src=\\"../img/logoFormLoveBook.png\\" class=\\"acount-logo\\" alt=\\"LoveBook logo\\">
</div>
<div class=\\"log-content\\">
    <form action=\\"\\" id=\\"login-form\\" class=\\"form\\">
        <label for=\\"login-email\\">Email</label>
        <input type=\\"email\\" id=\\"loginEmail\\" required=\\"\\">
        <label for=\\"login-password\\">Password</label>
        <input type=\\"password\\" id=\\"loginPassword\\" required=\\"\\">
        <div class=\\"error-message\\" id=\\"loginMessages\\"></div>
        <button type=\\"submit\\" id=\\"btnLogin\\">Login</button>
        <div class=\\"division\\">
            <hr class=\\"division-1\\">
            <p class=\\"division-2\\">O</p>
            <hr class=\\"division-3\\">
        </div>
        <div class=\\"login-google\\">
            <img src=\\"../img/icongoogle.png\\" alt=\\"logo google\\" class=\\"logo-google\\">
            <a id=\\"withGoogle\\">Login with google</a>
        </div>
        <div class=\\"pieform\\">
            <p style=\\"font-size: 12px;\\">You do not have an account</p>   
        </div>
        <button type=\\"submit\\" id=\\"signup\\">Sign Up</button>
    </form>
</div></div>"
`;

exports[`PostForm Should render 1`] = `
"<div><header class=\\"header-div\\">
    <nav>
    <div class=\\"logo-nav\\">
        <img class=\\"logo-nav\\" src=\\"../img/logo-nav2.png\\" alt=\\"\\">
    </div>
    </nav>
    <div class=\\"menu-nav\\">
        <ul>
        <li><a id=\\"logout\\" href=\\"\\"> <div class=\\"signOff\\"> <img class=\\"icon-nav\\" src=\\"../img/icons8-salir-redondeado-64.png\\"> <p>Sign off</p> </div> </a></li> 
            <!-- <li><a href=\\"#\\" id=\\"profilePerfil\\"><img class= \\"icon-nav\\" src=\\"../img/iconsusuario.png\\">Profile</a></li>-->
        </ul>
    </div>
    
</header><div class=\\"containerpost\\">
<div id=\\"addPublication\\">
    <div class=\\"form-Background\\">
    <form id=\\"publicationsForm\\" class=\\"form\\">
        <label for=\\"title\\">Title of the book</label>
        <input class=\\"inputpost\\" type=\\"text\\" id=\\"title\\" required=\\"\\">
        <label for=\\"rating\\">Rating</label>
        <input class=\\"inputpost\\" type=\\"number\\" id=\\"rating\\" min=\\"1\\" max=\\"10\\" required=\\"\\">
        <label for=\\"review\\">Review</label>
        <textarea class=\\"inputpost\\" type=\\"text\\" id=\\"review\\"></textarea>
        <div class=\\"error-formPost\\" id=\\"postMesseges\\"></div>
        <button class=\\"btnPost\\" id=\\"btnAddPoster\\" type=\\"submit\\"><img class=\\"icon-plus\\" src=\\"../img/plusazul.png\\">Add Post</button>    
    </form>
    </div>
</div>
</div></div>"
`;

exports[`SignUp Should render SignUp 1`] = `
"<div class=\\"acount\\">
        <div class=\\"acount-header\\">
            <img src=\\"../img/logoFormLoveBook.png\\" class=\\"acount-logo\\" alt=\\"LoveBook logo\\">
        </div>
        <div class=\\"log-content\\">
            <form action=\\"\\" id=\\"signupForm\\" class=\\"form\\">
                <label for=\\"signupUserName\\">Username</label>
                <input type=\\"text\\" id=\\"signupUserName\\">
                <label for=\\"signupEmail\\">Email</label>
                <input type=\\"email\\" id=\\"signupEmail\\" required=\\"\\">
                <label for=\\"signupPassword\\">Password</label>
                <input type=\\"password\\" id=\\"signupPassword\\" required=\\"\\">
                <label for=\\"signupPassword2\\">Confirm your password</label>
                <input type=\\"password\\" id=\\"signupPassword2\\" required=\\"\\">
                <div class=\\"error-message\\" id=\\"signupMesseges\\"></div>
                <button type=\\"submit\\" id=\\"btnSendSignUp\\">Sign Up</button>
            </form>
        </div>
        </div>"
`;
